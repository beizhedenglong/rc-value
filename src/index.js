import React from 'react'

const isNil = x => x == null

export const useValue = (props) => {
  const {
    defaultValue,
    value: parentValue,
    onChange = () => { },
  } = props
  const [value, set] = React.useState(!isNil(parentValue) ? parentValue : defaultValue)
  if (!isNil(parentValue) && (parentValue !== value)) {
    set(() => parentValue)
  }
  return {
    value,
    onChange: (updater, ...rest) => {
      set((prev) => {
        const newValue = typeof updater === 'function' ? updater(prev) : updater
        onChange(newValue, ...rest)
        return newValue
      })
    },
  }
}

export const Value = (props) => {
  const returned = useValue(props)
  return props.children(returned)
}

export const withValue = (BaseComponent) => {
  const WithValue = (props) => {
    const {
      defaultValue, value, onChange, ...rest
    } = props
    const returned = useValue(props)
    return <BaseComponent {...rest} {...returned} />
  }
  WithValue.displayName = `withValue(${BaseComponent.displayName})`
  return WithValue
}
