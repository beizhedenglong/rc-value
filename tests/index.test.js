import { render, fireEvent, cleanup } from 'react-testing-library'
import React from 'react'
import { Value, withValue, useValue } from '../src/index'

afterEach(cleanup)
describe('Value', () => {
  const Switch = props => (
    <Value {...props}>
      {({ value, onChange }) => (
        <button
          onClick={(e) => { e.persist(); onChange(prev => !prev, e) }}
          data-testid="button"
        >
          {`${value}`}
        </button>
      )}
    </Value>
  )
  test('Uncontrolled Component', () => {
    const onChangeMock = jest.fn()
    const {
      container,
      getByTestId,
    } = render(<Switch defaultValue={false} onChange={onChangeMock} />)
    expect(container).toMatchSnapshot()
    expect(getByTestId('button').textContent).toMatch(/false/)
    fireEvent.click(getByTestId('button'))
    expect(onChangeMock.mock.calls[0][0]).toBe(true)
    expect(onChangeMock.mock.calls[0][1]).toBeInstanceOf(Object)
    expect(getByTestId('button').textContent).toMatch(/true/)
  })
  test('Controlled Component', () => {
    const onChangeMock = jest.fn()
    const {
      container,
      getByTestId,
    } = render(<Switch value={false} onChange={onChangeMock} />)
    expect(container).toMatchSnapshot()
    expect(getByTestId('button').textContent).toMatch(/false/)
    fireEvent.click(getByTestId('button'))
    expect(onChangeMock.mock.calls[0][0]).toBe(true)
    expect(onChangeMock.mock.calls[0][1]).toBeInstanceOf(Object)
    expect(getByTestId('button').textContent).toMatch(/false/)
  })
})

describe('withValue', () => {
  const Switch = withValue(({ value, onChange }) => (
    <button
      onClick={(e) => { e.persist(); onChange(prev => !prev, e) }}
      data-testid="button"
    >
      {`${value}`}
    </button>
  ))
  test('Uncontrolled Component', () => {
    const onChangeMock = jest.fn()
    const {
      container,
      getByTestId,
    } = render(<Switch defaultValue={false} onChange={onChangeMock} />)
    expect(container).toMatchSnapshot()
    expect(getByTestId('button').textContent).toMatch(/false/)
    fireEvent.click(getByTestId('button'))
    expect(onChangeMock.mock.calls[0][0]).toBe(true)
    expect(onChangeMock.mock.calls[0][1]).toBeInstanceOf(Object)
    expect(getByTestId('button').textContent).toMatch(/true/)
  })
  test('Controlled Component', () => {
    const onChangeMock = jest.fn()
    const {
      container,
      getByTestId,
    } = render(<Switch value={false} onChange={onChangeMock} />)
    expect(container).toMatchSnapshot()
    expect(getByTestId('button').textContent).toMatch(/false/)
    fireEvent.click(getByTestId('button'))
    expect(onChangeMock.mock.calls[0][0]).toBe(true)
    expect(onChangeMock.mock.calls[0][1]).toBeInstanceOf(Object)
    expect(getByTestId('button').textContent).toMatch(/false/)
  })
})

describe('useValue', () => {
  const Switch = (props) => {
    const { value, onChange } = useValue(props)
    return (
      <button
        onClick={(e) => { e.persist(); onChange(prev => !prev, e) }}
        data-testid="button"
      >
        {`${value}`}
      </button>
    )
  }
  test('Uncontrolled Component', () => {
    const onChangeMock = jest.fn()
    const {
      container,
      getByTestId,
    } = render(<Switch defaultValue={false} onChange={onChangeMock} />)
    expect(container).toMatchSnapshot()
    expect(getByTestId('button').textContent).toMatch(/false/)
    fireEvent.click(getByTestId('button'))
    expect(onChangeMock.mock.calls[0][0]).toBe(true)
    expect(onChangeMock.mock.calls[0][1]).toBeInstanceOf(Object)
    expect(getByTestId('button').textContent).toMatch(/true/)
  })
  test('Controlled Component', () => {
    const onChangeMock = jest.fn()
    const {
      container,
      getByTestId,
    } = render(<Switch value={false} onChange={onChangeMock} />)
    expect(container).toMatchSnapshot()
    expect(getByTestId('button').textContent).toMatch(/false/)
    fireEvent.click(getByTestId('button'))
    expect(onChangeMock.mock.calls[0][0]).toBe(true)
    expect(onChangeMock.mock.calls[0][1]).toBeInstanceOf(Object)
    expect(getByTestId('button').textContent).toMatch(/false/)
  })
})
