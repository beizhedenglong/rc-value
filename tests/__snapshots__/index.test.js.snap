// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value Controlled Component 1`] = `
<div>
  <button
    data-testid="button"
  >
    false
  </button>
</div>
`;

exports[`Value Uncontrolled Component 1`] = `
<div>
  <button
    data-testid="button"
  >
    false
  </button>
</div>
`;

exports[`useValue Controlled Component 1`] = `
<div>
  <button
    data-testid="button"
  >
    false
  </button>
</div>
`;

exports[`useValue Uncontrolled Component 1`] = `
<div>
  <button
    data-testid="button"
  >
    false
  </button>
</div>
`;

exports[`withValue Controlled Component 1`] = `
<div>
  <button
    data-testid="button"
  >
    false
  </button>
</div>
`;

exports[`withValue Uncontrolled Component 1`] = `
<div>
  <button
    data-testid="button"
  >
    false
  </button>
</div>
`;
